d/bash

```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+docker-compose run --rm decidim bash "$@"
```

Comment:
Not sure `docker-compose` run is the best way to do this. I usually just launch the app via `docker-compose up -d` (this starts postgresql & potentially any other dependency) and then execute the commands via `docker-compose exec decidim whatever_command`. This boots faster and is more efficient memory-wise.

Reply:
The thing is... the `decidim` repo is not an app, so I'm not sure if `docker-compose up` means anything here... According to the current `docker-compose.yml` file, it would mean to generate a decidim application via the `decidim` executable (since it has `decidim` configured as the command for the `decidim` service). But I'm not really using it this way.

To boot a development application, I use the `rails` binstub added here. For the rest of the commands, I use the same command as in normal development, for example`, `d/rake test_app` to generate a test application, `d/rake development_app` to generate a development application, `d/rake test_all` to run all tests, `d/bundle install` to install dependencies, and so on.

Also, I'm not sure about the speed and memory efficiency argument since I haven't noticed any problem with it, it works more or less like regular development for me. So... I don't know, maybe there are better ways to do it, yet, I just haven't researched since it hasn't bothered me.

Finally, I don't really use this specific binstub a lot except for troubleshooting things inside the image. But I thought it could be provided for completeness so the user is able to run arbitrary commands inside the image.

Reply:
@deivid-rodriguez but you'll need postgres and other dependencies running when you generate a new app or test, right? Using `docker run whatever` will start postgres and other dependencies every time you run the command - by using `docker exec` you execute commands *inside* existing containers. That's why I say you'll cut boot time and you'll save memory :). Or maybe I'm missing something?

Reply:
@josepjaume I understand what you're saying: we don't need to start postgreSQL everytime you run a docker command, you can do that upfront. What I'm saying is that it doesn't matter to me to start postgreSQL every time since I haven't noticed the degradation, and making it work the way you suggest needs work. So, I'm totally up for doing it the way you like, but I think this is a good start since it works.

However, I'm doing some tests and I'm actually not sure about what you're saying... It seems to me that postgreSQL is booted just once...

```console
$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
06dc890c4cff        redis               "docker-entrypoint.s…"   2 weeks ago         Up About a minute   6379/tcp            decidim_redis_1
6b745e5319a7        postgres            "docker-entrypoint.s…"   2 weeks ago         Up About a minute   5432/tcp            decidim_pg_1

$ docker stop 06dc890c4cff 6b745e5319a7
06dc890c4cff
6b745e5319a7

$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

$ time docker-compose run --rm decidim ls
Starting decidim_pg_1 ... 
Starting decidim_pg_1
Starting decidim_redis_1 ... 
Starting decidim_redis_1 ... done
+ stat -c %u /code/Gemfile
+ USER_UID=1000
+ stat -c %g /code/Gemfile
+ USER_GID=1000
+ export USER_UID
+ export USER_GID
+ usermod -u 1000 decidim
+ groupmod -g 1000 decidim
+ usermod -g 1000 decidim
+ chown -R -h 1000 /usr/local/bundle
+ chgrp -R -h 1000 /usr/local/bundle
+ /usr/bin/sudo -EH -u decidim ls
CHANGELOG.md	    README.md	  d			  decidim-core			   decidim-surveys	  docs		 spec		      yarn.lock
CODE_OF_CONDUCT.md  Rakefile	  decidim-accountability  decidim-debates		   decidim-system	  jsconfig.json  tsconfig.json
CONTRIBUTING.md     bin		  decidim-admin		  decidim-dev			   decidim-verifications  lib		 tslint.json
Dockerfile.design   codecov.yml   decidim-api		  decidim-meetings		   decidim.gemspec	  logo.svg	 webpack.config.js
Gemfile		    config	  decidim-assemblies	  decidim-pages			   decidim_app-design	  node_modules	 webpack.d.ts
Gemfile.lock	    coverage	  decidim-budgets	  decidim-participatory_processes  development_app	  package.json	 webpack.report.html
LICENSE-AGPLv3.txt  crowdin.yaml  decidim-comments	  decidim-proposals		   docker-compose.yml	  pkg		 yarn-error.log

real	0m4.263s
user	0m0.956s
sys	0m0.112s

$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
06dc890c4cff        redis               "docker-entrypoint.s…"   2 weeks ago         Up 6 seconds        6379/tcp            decidim_redis_1
6b745e5319a7        postgres            "docker-entrypoint.s…"   2 weeks ago         Up 7 seconds        5432/tcp            decidim_pg_1

$ time docker-compose run --rm decidim ls
Starting decidim_redis_1 ... done
Starting decidim_pg_1 ... done
+ stat -c %u /code/Gemfile
+ USER_UID=1000
+ stat -c %g /code/Gemfile
+ USER_GID=1000
+ export USER_UID
+ export USER_GID
+ usermod -u 1000 decidim
+ groupmod -g 1000 decidim
+ usermod -g 1000 decidim
+ chown -R -h 1000 /usr/local/bundle
+ chgrp -R -h 1000 /usr/local/bundle
+ /usr/bin/sudo -EH -u decidim ls
CHANGELOG.md	    README.md	  d			  decidim-core			   decidim-surveys	  docs		 spec		      yarn.lock
CODE_OF_CONDUCT.md  Rakefile	  decidim-accountability  decidim-debates		   decidim-system	  jsconfig.json  tsconfig.json
CONTRIBUTING.md     bin		  decidim-admin		  decidim-dev			   decidim-verifications  lib		 tslint.json
Dockerfile.design   codecov.yml   decidim-api		  decidim-meetings		   decidim.gemspec	  logo.svg	 webpack.config.js
Gemfile		    config	  decidim-assemblies	  decidim-pages			   decidim_app-design	  node_modules	 webpack.d.ts
Gemfile.lock	    coverage	  decidim-budgets	  decidim-participatory_processes  development_app	  package.json	 webpack.report.html
LICENSE-AGPLv3.txt  crowdin.yaml  decidim-comments	  decidim-proposals		   docker-compose.yml	  pkg		 yarn-error.log

real	0m2.506s
user	0m0.808s
sys	0m0.100s
```

Reply:
Oh! That's weird. Maybe I don't know enough about docker... Let's merge this and iterate over it, if we need to. Thanks!

Reply:
Thanks @josepjaume, we'll learn more as we go!